### DISCUSSION TITLE
获取CMake库的其他导出变量

### DISCUSSION THREAD

**Discussion by Smalldy** [xmake-io/xmake#2478] [Q&A]
xmake可以自动的链接cmake提供的库，但是有些cmake库可能还会导出一些其他的变量作为他用，我举个例子，你引用了一个TESTCMAKELIB库，在你自己的cmake中你可能需要 
`add_definitions(${TESTCMAKELIB_DEFINITIONS})`这样一个步骤。如果在xmake中，我如何获得`TESTCAMELIB_DEFINITIONS`这个变量呢？

**Comment by waruqi** [comment]
通常你得自己 add_defines() 下，在你的 xmake.lua 里面，如果走的 package 集成，那么 package 里面，add_defines  自己配置下

**Reply by Smalldy** [reply]
我只知道要 需要 add_definitions(${TESTCAMELIB_DEFINITIONS}) TESTCAMELIB_DEFINITIONS 是cmake提供的 我并不知道其中的具体内容，

**Reply by waruqi** [reply]
这些不都是 cmake 内部编译 lib 用的么，为啥还会导出？ 对应的头文件引用时候，也需要定义这个？如果是对外也要导出，按理安装的 cmake import files 里面，也会有才对，你先确认下。。

**Reply by Smalldy** [reply]
确实是在使用xx库的时候需要 add_definitions，其实不是个 add_definitions的问题，而是个访问变量的问题
干说有些抽象
实际的情况大概长这个样子（做了处理）
```
find_package(TESTLIB "6.0.1" REQUIRED
  COMPONENTS a b c d )

add_definitions(${TESTLIB_DEFINITIONS})  <<-- 这里如何用xmake访问到 TESTLIB_DEFINITIONS

message(${PROJECT_SOURCE_DIR})
include_directories(
  ${TESTLIB_INCLUDE_DIRS}
  ${PROJECT_SOURCE_DIR}xxx/include
)

set(FRAMEWORK_TEST_LIBS
    TEST::c_api
    TEST::cpp_api
    TEST::cpp2_api
    ....
    ${TEST_EXTERNAL_LIBS} <<-----这里如何用xmake访问到 TEST_EXTERNAL_LIBS
)

......
```

**Reply by waruqi** [reply]
没懂，这是 cmake 里面的变量，跟 xmake 有啥关系，干嘛要 xmake 访问到。。你现在怎么在 xmake 里面用的 cmake库。。

**Reply by Smalldy** [reply]
>If we integrate and find cmake dependent packages in the xmake.lua project configuration, we usually don't need to use find_package directly, and we can use a more general and simple package integration method.
```cpp
add_requires("cmake::ZLIB", {alias = "zlib", system = true})
target("test")
    set_kind("binary")
    add_files("src/*.c")
    add_packages("zlib")
```
>We specify system = true to tell xmake to force cmake to find the package from the system. If it cannot be found, the installation logic will not be followed, because cmake does not provide the installation function of package managers such as vcpkg/conan. Only the package search feature is provided.

当 add_packages 不能完全引用此库时 可能是需要进行额外配置 例如 add_definitions ，因此我提出问题 如何在xmake中访问cmake的导出变量，希望通过xmake的add_defines() 将必要的东西引进来。

**Reply by waruqi** [reply]
如果你走 cmake::xxx 集成的，你可以调下 https://github.com/xmake-io/xmake/blob/master/xmake/modules/package/manager/cmake/find_package.lua 这个地方，按理里面有对 defines 的解析提取，但也没特别确认过，可以看下哪里遗漏了，导致 defines 没被提取到。。

你可以加点 print 调下，然后提个 pr 过来

另外，建议尽量走 add_requires("zlib")，更加易用，兼容性和跨平台性更好
